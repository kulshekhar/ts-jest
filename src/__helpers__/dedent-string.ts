export const omitLeadingWhitespace = (text: string): string => {
  return text.replace(/^\s+/gm, '')
}

export const dedent = (strings: TemplateStringsArray, ...values: unknown[]) => {
  let joinedString = ''
  for (let i = 0; i < values.length; i++) {
    joinedString += `${strings[i]}${values[i]}`
  }
  joinedString += strings[strings.length - 1]

  return omitLeadingWhitespace(joinedString)
}
