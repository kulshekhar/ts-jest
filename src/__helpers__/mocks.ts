import { testing } from 'bs-logger'

import { rootLogger } from '../util/logger'

export const logTargetMock = () => (rootLogger as testing.LoggerMock).target

export const mockObject = <T, M>(obj: T, newProps: M): T & M & { mockRestore: () => T } => {
  const backup: any = Object.create(null)

  Object.keys(newProps).forEach(key => {
    const desc = (backup[key] = Object.getOwnPropertyDescriptor(obj, key))
    const newDesc: any = { ...desc }
    if (newDesc.get) {
      newDesc.get = () => (newProps as any)[key]
    } else {
      newDesc.value = (newProps as any)[key]
    }
    Object.defineProperty(obj, key, newDesc)
  })
  if ((obj as any).mockRestore) backup.mockRestore = Object.getOwnPropertyDescriptor(obj, 'mockRestore')
  return Object.defineProperty(obj, 'mockRestore', {
    value() {
      Object.keys(backup).forEach(key => {
        Object.defineProperty(obj, key, backup[key])
      })
      return obj
    },
    configurable: true,
  })
}

export const mockWriteStream = () => ({
  written: [] as string[],
  write(text: string) {
    this.written.push(text)
  },
  clear() {
    this.written = []
  },
})
