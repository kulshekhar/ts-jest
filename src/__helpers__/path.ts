import { ensureDirSync } from 'fs-extra'
import { tmpdir } from 'os'
import { join, resolve } from 'path'

export const ROOT = resolve(__dirname, '..', '..')

export function tempDir(ns: string): string {
  const dir = join(tmpdir(), 'ts-jest-unit-tests', ns, parseInt((Math.random() * 1e17) as any, 10).toString(36))
  ensureDirSync(dir)

  return dir
}
