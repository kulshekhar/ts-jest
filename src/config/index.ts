export * from './config-set'
export * from './paths-to-module-name-mapper'
