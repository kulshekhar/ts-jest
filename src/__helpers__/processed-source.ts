import { isAbsolute, relative } from 'path'
import { RawSourceMap } from 'source-map'

import { ROOT } from './path'
import { ParsedSourceWithMaps, parseSource, relativisePaths, rewriteSourceMaps } from './source-maps'

export default class ProcessedSource {
  readonly filename: string

  constructor(readonly output: string, filename: string, readonly cwd: string = ROOT) {
    if (isAbsolute(filename)) filename = relative(this.cwd, filename)
    this.filename = filename
  }
  get outputCode(): string {
    return this.output
  }
  get parsedSource(): ParsedSourceWithMaps {
    return parseSource(this.output)
  }
  get outputCodeWithoutMaps(): string {
    return this.parsedSource.source
  }
  get outputSourceMaps(): RawSourceMap | undefined {
    return this.parsedSource.sourceMaps
  }
  get normalizedOutputCode(): string {
    return rewriteSourceMaps(this.output, this.sourceMapsNormalizer)
  }
  get normalizedOutputSourceMaps(): RawSourceMap | undefined {
    const maps = this.outputSourceMaps
    if (maps) return this.sourceMapsNormalizer(maps)

    return undefined
  }
  get sourceMapsNormalizer() {
    return (maps: RawSourceMap): RawSourceMap => relativisePaths(maps, this.cwd, '<cwd>/')
  }
}
