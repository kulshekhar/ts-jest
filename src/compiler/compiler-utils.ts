import { Logger } from 'bs-logger'
import { writeFileSync } from 'fs'
import micromatch = require('micromatch')
import { dirname, join, normalize, relative, resolve } from 'path'
import * as _ts from 'typescript'

import { ConfigSet } from '../config/config-set'
import { EXTENSION_REGEX, JSON_REGEX, TS_TSX_REGEX } from '../constants'
import { MemoryCache, SourceOutput, TSFiles } from '../types'
import { sha1 } from '../util/sha1'

/**
 * @internal
 */
export function getResolvedModulesCache(cacheDir: string): string {
  return join(cacheDir, sha1('ts-jest-resolved-modules', '\x00'))
}

/**
 * @internal
 * Get resolved modules of a test file and put into memory cache
 */
/* istanbul ignore next (we leave this for e2e) */
export function cacheResolvedModules(
  fileName: string,
  fileContent: string,
  memoryCache: MemoryCache,
  program: _ts.Program,
  cacheDir: string,
  logger: Logger,
) {
  // @ts-ignore
  const importReferences = program.getSourceFile(fileName)!.imports
  /**
   * Ugly trick while waiting for https://github.com/microsoft/TypeScript/issues/33994
   */
  if (importReferences.length) {
    logger.debug({ fileName }, 'cacheResolvedModules(): get resolved modules')

    memoryCache.resolvedModules[fileName] = Object.create(null)
    memoryCache.resolvedModules[fileName].modulePaths = importReferences
      .filter((importReference: any) => importReference.parent.parent.resolvedModules?.get(importReference.text))
      .map((importReference: any) =>
        normalize(
          (importReference.parent.parent.resolvedModules.get(importReference.text) as _ts.ResolvedModule)
            .resolvedFileName,
        ),
      )
      .reduce((a: any, b: any) => a.concat(b), [])
    memoryCache.resolvedModules[fileName].testFileContent = fileContent
    writeFileSync(getResolvedModulesCache(cacheDir), JSON.stringify(memoryCache.resolvedModules))
  }
}

/**
 * @internal
 */
export function isTestFile(testMatchPatterns: (string | RegExp)[], fileName: string) {
  return testMatchPatterns.some(pattern =>
    typeof pattern === 'string' ? micromatch.isMatch(fileName, pattern) : pattern.test(fileName),
  )
}

/* istanbul ignore next (we leave this for e2e) */
function isUsingProjectReferences(
  program: _ts.Program,
  projectReferences: readonly _ts.ProjectReference[] | undefined,
) {
  if (projectReferences && !!program.getProjectReferences) {
    return Boolean(program && program.getProjectReferences())
  }

  return false
}

/* istanbul ignore next (we leave this for e2e) */
function getResolvedProjectReferences(
  program: _ts.Program,
): readonly (_ts.ResolvedProjectReference | undefined)[] | undefined {
  const getProjectReferences = program.getResolvedProjectReferences ?? program.getProjectReferences
  if (getProjectReferences) {
    return getProjectReferences()
  }

  return
}

/* istanbul ignore next (we leave this for e2e) */
function getProjectReferenceForFile(
  filePath: string,
  program: _ts.Program,
  projectReferences: readonly _ts.ProjectReference[] | undefined,
) {
  if (isUsingProjectReferences(program, projectReferences)) {
    return (
      program &&
      getResolvedProjectReferences(program)!.find(
        ref => (ref && ref.commandLine.fileNames.some(file => normalize(file) === filePath)) || false,
      )
    )
  }

  return
}

/**
 * @internal
 */
/* istanbul ignore next (we leave this for e2e) */
export function getAndCacheProjectReference(
  filePath: string,
  program: _ts.Program,
  files: TSFiles,
  projectReferences: readonly _ts.ProjectReference[] | undefined,
) {
  const file = files.get(filePath)
  if (file?.projectReference) {
    return file.projectReference.project
  }

  const projectReference = getProjectReferenceForFile(filePath, program, projectReferences)
  if (file !== undefined) {
    file.projectReference = { project: projectReference }
  }

  return projectReference
}

// Adapted from https://github.com/Microsoft/TypeScript/blob/45101491c0b077c509b25830ef0ee5f85b293754/src/compiler/tsbuild.ts#L305
/* istanbul ignore next (we leave this for e2e) */
function getOutputJavaScriptFileName(inputFileName: string, projectReference: _ts.ResolvedProjectReference) {
  const { options } = projectReference.commandLine
  const projectDirectory = options.rootDir || dirname(projectReference.sourceFile.fileName)
  const relativePath = relative(projectDirectory, inputFileName)
  const outputPath = resolve(options.outDir || projectDirectory, relativePath)
  const newExtension = JSON_REGEX.test(inputFileName)
    ? '.json'
    : TS_TSX_REGEX.test(inputFileName) && options.jsx === _ts.JsxEmit.Preserve
    ? '.jsx'
    : '.js'

  return outputPath.replace(EXTENSION_REGEX, newExtension)
}

/**
 * Gets the output JS file path for an input file governed by a composite project.
 * Pulls from the cache if it exists; computes and caches the result otherwise.
 */
/* istanbul ignore next (we leave this for e2e) */
function getAndCacheOutputJSFileName(
  inputFileName: string,
  projectReference: _ts.ResolvedProjectReference,
  files: TSFiles,
) {
  const file = files.get(inputFileName)
  if (file?.projectReference?.outputFileName) {
    return file.projectReference.outputFileName
  }

  const outputFileName = getOutputJavaScriptFileName(inputFileName, projectReference)
  if (file !== undefined) {
    file.projectReference = file.projectReference ?? {
      project: projectReference,
    }
    file.projectReference.outputFileName = outputFileName
  }

  return outputFileName
}

/**
 * @internal
 */
/* istanbul ignore next (we leave this for e2e) */
export function getCompileResultFromReferencedProject(
  fileName: string,
  configs: ConfigSet,
  files: TSFiles,
  referencedProject: _ts.ResolvedProjectReference,
): SourceOutput {
  const [relativeProjectConfigPath, relativeFilePath] = [
    configs.resolvePath(referencedProject.sourceFile.fileName),
    configs.resolvePath(fileName),
  ]
  if (referencedProject.commandLine.options.outFile !== undefined) {
    throw new Error(
      `The referenced project at ${relativeProjectConfigPath} is using ` +
        "the outFile' option, which is not supported with ts-jest.",
    )
  }

  const jsFileName = getAndCacheOutputJSFileName(fileName, referencedProject, files)
  const relativeJSFileName = configs.resolvePath(jsFileName)
  if (!configs.compilerModule.sys.fileExists(jsFileName)) {
    throw new Error(
      'Could not find output JavaScript file for input ' +
        `${relativeFilePath} (looked at ${relativeJSFileName}).\n` +
        'The input file is part of a project reference located at ' +
        `${relativeProjectConfigPath}, so ts-jest is looking for the ` +
        'projectâ€™s pre-built output on disk. Try running `tsc --build` ' +
        'to build project references.',
    )
  }

  const mapFileName = `${jsFileName}.map`
  const outputText = configs.compilerModule.sys.readFile(jsFileName)
  const sourceMapText = configs.compilerModule.sys.readFile(mapFileName)

  return [outputText!, sourceMapText!]
}
