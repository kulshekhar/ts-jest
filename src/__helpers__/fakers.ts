import { Config } from '@jest/types'
import { resolve } from 'path'

import { createCompilerInstance } from '../compiler/instance'
import { ConfigSet } from '../config/config-set'
import { BabelConfig, TsCompiler, TsJestConfig, TsJestGlobalOptions } from '../types'
import { ImportReasons } from '../util/messages'

export function filePath(relPath: string): string {
  return resolve(__dirname, '..', '..', relPath)
}

export const rootDir = filePath('')

export function tsJestConfig(options?: Partial<TsJestConfig>): TsJestConfig {
  return {
    isolatedModules: false,
    compiler: 'typescript',
    transformers: [],
    babelConfig: undefined,
    tsConfig: undefined,
    packageJson: undefined,
    stringifyContentPathRegex: undefined,
    diagnostics: { ignoreCodes: [], pretty: false, throws: true },
    ...options,
  }
}

function getJestConfig<T extends Config.ProjectConfig>(
  options?: Partial<Config.InitialOptions | Config.ProjectConfig>,
  tsJestOptions?: TsJestGlobalOptions,
): T {
  const res = {
    globals: {},
    ...options,
  } as any
  if (tsJestOptions) {
    res.globals['ts-jest'] = tsJestOptions
  }

  return res
}

export function babelConfig<T extends BabelConfig>(options?: BabelConfig): T {
  return {
    ...options,
    presets: [...(options && options.presets)],
    plugins: [...(options && options.plugins)],
  } as any
}

export function importReason(text = '[[BECAUSE]]'): ImportReasons {
  return text as any
}

export const defaultResolve = (path: string): string => `resolved:${path}`

export function createConfigSet({
  jestConfig,
  tsJestConfig,
  parentConfig,
  resolve = defaultResolve,
  ...others
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
  resolve?: ((path: string) => string) | null
  [key: string]: any
} = {}): ConfigSet {
  const defaultTestRegex = ['(/__tests__/.*|(\\\\.|/)(test|spec))\\\\.[jt]sx?$']
  const defaultTestMatch = ['**/__tests__/**/*.[jt]s?(x)', '**/?(*.)+(spec|test).[jt]s?(x)']
  jestConfig = {
    ...jestConfig,
    testMatch: jestConfig?.testMatch ? [...jestConfig.testMatch, ...defaultTestMatch] : defaultTestMatch,
    testRegex: jestConfig?.testRegex ? [...defaultTestRegex, ...jestConfig.testRegex] : defaultTestRegex,
  }
  const cs = new ConfigSet(getJestConfig(jestConfig, tsJestConfig), parentConfig)
  if (resolve) {
    cs.resolvePath = resolve
  }
  Object.keys(others).forEach((key) => {
    Object.defineProperty(cs, key, { value: others[key] })
  })

  return cs
}

// not really unit-testing here, but it's hard to mock all those values :-D
export function makeCompiler({
  jestConfig,
  tsJestConfig,
  parentConfig,
}: {
  jestConfig?: Partial<Config.ProjectConfig>
  tsJestConfig?: TsJestGlobalOptions
  parentConfig?: TsJestGlobalOptions
} = {}): TsCompiler {
  tsJestConfig = { ...tsJestConfig }
  tsJestConfig.diagnostics = {
    ...(tsJestConfig.diagnostics as any),
    pretty: false,
  }
  const cs = createConfigSet({ jestConfig, tsJestConfig, parentConfig, resolve: null })

  return createCompilerInstance(cs)
}
