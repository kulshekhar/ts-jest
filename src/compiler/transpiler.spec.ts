import { LogLevels } from 'bs-logger'
import { removeSync, writeFileSync } from 'fs-extra'
import * as _ts from 'typescript'

import { makeCompiler } from '../__helpers__/fakers'
import { logTargetMock } from '../__helpers__/mocks'
import ProcessedSource from '../__helpers__/processed-source'
import { TS_JEST_OUT_DIR } from '../config/config-set'

import * as compilerUtils from './compiler-utils'

const logTarget = logTargetMock()

describe('Transpiler', () => {
  const baseTsJestConfig = {
    isolatedModules: true,
  }

  beforeEach(() => {
    logTarget.clear()
  })

  it('should compile using transpileModule and not use cache', () => {
    const compiler = makeCompiler({ tsJestConfig: { ...baseTsJestConfig, tsConfig: false } })
    const spy = jest.spyOn(_ts, 'transpileModule')

    logTarget.clear()
    const compiled = compiler.compile('export default 42', __filename)

    expect(new ProcessedSource(compiled, __filename)).toMatchSnapshot()
    expect(spy).toHaveBeenCalled()
    expect(logTarget.filteredLines(LogLevels.debug, Infinity)).toMatchInlineSnapshot(`
      Array [
        "[level:20] readThrough(): no cache
      ",
        "[level:20] compileFn(): compiling as isolated module
      ",
        "[level:20] visitSourceFileNode(): hoisting
      ",
      ]
    `)

    spy.mockRestore()
  })

  it(
    'should call createProgram() with projectReferences, call getAndCacheProjectReference()' +
      ' and getCompileResultFromReferenceProject() when there are projectReferences from tsconfig',
    () => {
      const programSpy = jest.spyOn(_ts, 'createProgram')
      const source = 'console.log("hello")'
      const fileName = 'isolated-test-reference-project.ts'
      const getAndCacheProjectReferenceSpy = jest
        .spyOn(compilerUtils, 'getAndCacheProjectReference')
        .mockReturnValueOnce({} as any)
      jest
        .spyOn(compilerUtils, 'getCompileResultFromReferencedProject')
        .mockImplementationOnce(() => [
          source,
          '{"version":3,"file":"isolated-test-reference-project.js","sourceRoot":"","sources":["isolated-test-reference-project.ts"],"names":[],"mappings":"AAAA,OAAO,CAAC,GAAG,CAAC,OAAO,CAAC,CAAA","sourcesContent":["console.log(\\"hello\\")"]}',
        ])
      writeFileSync(fileName, source)
      const compiler = makeCompiler({
        tsJestConfig: {
          ...baseTsJestConfig,
          tsConfig: 'src/__mocks__/tsconfig-project-references.json',
        },
      })
      compiler.compile(source, fileName)

      expect(programSpy).toHaveBeenCalled()
      expect((programSpy.mock.calls[0][0] as any).options.configFilePath).toContain('tsconfig-project-references.json')
      expect(getAndCacheProjectReferenceSpy).toHaveBeenCalled()
      expect(compilerUtils.getCompileResultFromReferencedProject).toHaveBeenCalled()

      jest.restoreAllMocks()
      removeSync(fileName)
    },
  )

  it('should call createProgram() without projectReferences when there are no projectReferences from tsconfig', () => {
    const programSpy = jest.spyOn(_ts, 'createProgram')
    const source = 'console.log("hello")'
    const fileName = 'isolated-test-reference-project-1.ts'
    const getAndCacheProjectReferenceSpy = jest
      .spyOn(compilerUtils, 'getAndCacheProjectReference')
      .mockReturnValueOnce(undefined)
    jest.spyOn(compilerUtils, 'getCompileResultFromReferencedProject')
    writeFileSync(fileName, source, 'utf8')
    const compiler = makeCompiler({
      tsJestConfig: {
        ...baseTsJestConfig,
        tsConfig: false,
      },
    })
    compiler.compile(source, fileName)

    expect(programSpy).toHaveBeenCalled()
    expect((programSpy.mock.calls[0][1] as any).configFilePath).toBeUndefined()
    expect(getAndCacheProjectReferenceSpy).toHaveBeenCalled()
    expect(compilerUtils.getCompileResultFromReferencedProject).not.toHaveBeenCalled()

    jest.restoreAllMocks()
    removeSync(fileName)
  })

  it('should compile js file for allowJs true', () => {
    const fileName = `${__filename}.test.js`
    const compiler = makeCompiler({
      tsJestConfig: { ...baseTsJestConfig, tsConfig: { allowJs: true, outDir: TS_JEST_OUT_DIR } },
    })
    const source = 'export default 42'

    writeFileSync(fileName, source, 'utf8')
    const compiled = compiler.compile(source, fileName)

    expect(new ProcessedSource(compiled, fileName)).toMatchSnapshot()

    removeSync(fileName)
  })

  it('should compile tsx file for jsx preserve', () => {
    const fileName = `foo.tsx`
    const compiler = makeCompiler({
      tsJestConfig: {
        ...baseTsJestConfig,
        tsConfig: {
          jsx: 'preserve' as any,
        },
      },
    })
    const source = `
        const App = () => {
          return <>Test</>
        }
      `

    writeFileSync(fileName, source, 'utf8')
    const compiled = compiler.compile(source, fileName)

    expect(new ProcessedSource(compiled, fileName)).toMatchSnapshot()

    removeSync(fileName)
  })

  it('should compile tsx file for other jsx options', () => {
    const fileName = `foo.tsx`
    const compiler = makeCompiler({
      tsJestConfig: {
        ...baseTsJestConfig,
        tsConfig: {
          jsx: 'react' as any,
        },
      },
    })
    const source = `
        const App = () => {
          return <>Test</>
        }
      `

    writeFileSync(fileName, source, 'utf8')
    const compiled = compiler.compile(source, fileName)

    expect(new ProcessedSource(compiled, fileName)).toMatchSnapshot()

    removeSync(fileName)
  })

  it('should have correct source maps', () => {
    const compiler = makeCompiler({ tsJestConfig: { ...baseTsJestConfig, tsConfig: false } })
    const source = 'const f = (v: number) => v\nconst t: number = f(5)'

    const compiled = compiler.compile(source, __filename)

    expect(new ProcessedSource(compiled, __filename).outputSourceMaps).toMatchObject({
      file: __filename,
      sources: [__filename],
      sourcesContent: [source],
    })
  })

  it('should not report diagnostics related to typings', () => {
    const compiler = makeCompiler({ tsJestConfig: { ...baseTsJestConfig, tsConfig: false } })

    expect(() =>
      compiler.compile(
        `
const f = (v: number) => v
const t: string = f(5)
const v: boolean = t
`,
        'foo.ts',
      ),
    ).not.toThrowError()
  })

  it('should report diagnostics related to codes with pathRegex config is undefined', () => {
    const compiler = makeCompiler({ tsJestConfig: { ...baseTsJestConfig, tsConfig: false } })

    expect(() =>
      compiler.compile(
        `
const f = (v: number) = v
const t: string = f(5)
`,
        'foo.ts',
      ),
    ).toThrowErrorMatchingSnapshot()
  })

  it('should report diagnostics related to codes with pathRegex config matches file name', () => {
    const compiler = makeCompiler({
      tsJestConfig: { ...baseTsJestConfig, tsConfig: false, diagnostics: { pathRegex: 'foo.ts' } },
    })

    expect(() =>
      compiler.compile(
        `
const f = (v: number) = v
const t: string = f(5)
`,
        'foo.ts',
      ),
    ).toThrowErrorMatchingSnapshot()
  })

  it('should not report diagnostics related to codes with pathRegex config does not match file name', () => {
    const compiler = makeCompiler({
      tsJestConfig: { ...baseTsJestConfig, tsConfig: false, diagnostics: { pathRegex: 'bar.ts' } },
    })

    expect(() =>
      compiler.compile(
        `
const f = (v: number) = v
const t: string = f(5)
`,
        'foo.ts',
      ),
    ).not.toThrowError()
  })
})
