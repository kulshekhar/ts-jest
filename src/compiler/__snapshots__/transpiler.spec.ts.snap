// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Transpiler diagnostics should report diagnostics related to codes with pathRegex config is undefined 1`] = `"foo.ts(2,23): error TS1005: '=>' expected."`;

exports[`Transpiler diagnostics should report diagnostics related to codes with pathRegex config matches file name 1`] = `"foo.ts(2,23): error TS1005: '=>' expected."`;

exports[`Transpiler jsx option should compile tsx file for jsx preserve 1`] = `
  ===[ FILE: foo.tsx ]============================================================
  "use strict";
  var App = function () {
      return <>Test</>;
  };
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiZm9vLnRzeCIsIm1hcHBpbmdzIjoiO0FBQ1EsSUFBTSxHQUFHLEdBQUc7SUFDVixPQUFPLEVBQUUsSUFBSSxHQUFHLENBQUE7QUFDbEIsQ0FBQyxDQUFBIiwibmFtZXMiOltdLCJzb3VyY2VzIjpbImZvby50c3giXSwic291cmNlc0NvbnRlbnQiOlsiXG4gICAgICAgIGNvbnN0IEFwcCA9ICgpID0+IHtcbiAgICAgICAgICByZXR1cm4gPD5UZXN0PC8+XG4gICAgICAgIH1cbiAgICAgICJdLCJ2ZXJzaW9uIjozfQ==
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: foo.tsx
  mappings: ';AACQ,IAAM,GAAG,GAAG;IACV,OAAO,EAAE,IAAI,GAAG,CAAA;AAClB,CAAC,CAAA'
  names: []
  sources:
    - foo.tsx
  sourcesContent:
    - |2-
  
              const App = () => {
                return <>Test</>
              }
            
  version: 3
  ================================================================================
`;

exports[`Transpiler jsx option should compile tsx file for other jsx options 1`] = `
  ===[ FILE: foo.tsx ]============================================================
  "use strict";
  var App = function () {
      return React.createElement(React.Fragment, null, "Test");
  };
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiZm9vLnRzeCIsIm1hcHBpbmdzIjoiO0FBQ1EsSUFBTSxHQUFHLEdBQUc7SUFDVixPQUFPLGlEQUFTLENBQUE7QUFDbEIsQ0FBQyxDQUFBIiwibmFtZXMiOltdLCJzb3VyY2VzIjpbImZvby50c3giXSwic291cmNlc0NvbnRlbnQiOlsiXG4gICAgICAgIGNvbnN0IEFwcCA9ICgpID0+IHtcbiAgICAgICAgICByZXR1cm4gPD5UZXN0PC8+XG4gICAgICAgIH1cbiAgICAgICJdLCJ2ZXJzaW9uIjozfQ==
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: foo.tsx
  mappings: ';AACQ,IAAM,GAAG,GAAG;IACV,OAAO,iDAAS,CAAA;AAClB,CAAC,CAAA'
  names: []
  sources:
    - foo.tsx
  sourcesContent:
    - |2-
  
              const App = () => {
                return <>Test</>
              }
            
  version: 3
  ================================================================================
`;

exports[`Transpiler should compile js file for allowJs true 1`] = `
  ===[ FILE: foo.js ]=============================================================
  "use strict";
  Object.defineProperty(exports, "__esModule", { value: true });
  exports.default = 42;
  //# sourceMappingURL=data:application/json;charset=utf-8;base64,eyJmaWxlIjoiZm9vLmpzIiwibWFwcGluZ3MiOiI7O0FBQUEsa0JBQWUsRUFBRSxDQUFBIiwibmFtZXMiOltdLCJzb3VyY2VzIjpbImZvby5qcyJdLCJzb3VyY2VzQ29udGVudCI6WyJleHBvcnQgZGVmYXVsdCA0MiJdLCJ2ZXJzaW9uIjozfQ==
  ===[ INLINE SOURCE MAPS ]=======================================================
  file: foo.js
  mappings: ';;AAAA,kBAAe,EAAE,CAAA'
  names: []
  sources:
    - foo.js
  sourcesContent:
    - export default 42
  version: 3
  ================================================================================
`;
